### ISSUE TITLE
Scope-shifting issue using peek

### ISSUE THREAD

**Issue by sebastianvoss** [functional-streams-for-scala/fs2#1982]
I posted this also in Gitter and dicussed in `http4s` channel with @rossabaker but maybe it is easier to track/ follow up as an Github issue.

I have a program which reads data from JDBC (using `doobie`) and streams it via an `http4s` endpoint. When my SQL throws an error I would like to send a non-200 response.

To identify SQL errors, I have to start consuming the stream. Therefore I thought to use `peek` and check if the stream is returning at least one chunk. The problem with this approach is that the `doobie` stream gets finalized and I cannot consume the ResultSet anymore (I get `PSQLException: This ResultSet is closed`).

Here is a sample snippet (I tried to remove dependencies, hopefully it makes it easier to reproduce it):

```scala
object Test extends IOApp {
  override def run(args: List[String]): IO[ExitCode] = {
    val s = for {
      closed <- Stream.eval(Ref.of[IO, Boolean](false))
      stream <- Stream.eval(stream(result(closed)))
      _      <- Stream.eval(closed.get.flatMap(b => IO(println(b))))  // <- prints true
      _      <- stream.evalTap(s => IO(println(s))) // <- in real-world this then fails with ResultSet closed exception
    } yield ()

    s.compile.drain.as(ExitCode.Success)
  }

  private def result(closed: Ref[IO, Boolean]): Stream[IO, String] =
    Stream
      .emits[IO, String]((1 to 10).map(_.toString))
      .onFinalize(closed.set(true))

  private def stream(result: Stream[IO, String]): IO[Stream[IO, String]] =
    result.pull.peek
      .flatMap {
        case Some((_, s)) => Pull.output1(IO(Stream.emit("ok") ++ s))
        case None         => Pull.output1(IO(Stream.emit("error")))
      }
      .stream
      .compile
      .lastOrError
      .flatten
}
```

**Comment by mpilquist** [comment]
The `stream` method as written in this example (intentionally) leaks a reference to the tail of the stream provided by `peek`. As a result of compiling `result`, all resources are freed at the completion of evaluation of the compiled `IO`. Here's an annotated version that attempts to clarify that:

```scala
  private def stream(result: Stream[IO, String]): IO[Stream[IO, String]] =
    result.pull.peek
      .flatMap {
        case Some((_, s)) => Pull.output1(IO(Stream.emit("ok") ++ s)) // <-- s is emitted here
        case None         => Pull.output1(IO(Stream.emit("error")))
      }
      .stream
      .compile
      .lastOrError // <-- s is returned here
      .flatten // <-- when this IO is evaluated, it releases all stream resources before returning the result, closing whatever resources were used in the production of the `result` stream
```

To correct this, you have to avoid compiling the stream twice.

**Comment by sebastianvoss** [comment]
@mpilquist Thanks for the clarification. That makes perfect sense. Though, I'm currently out of ideas how to avoid compiling it twice and still reach my goal. In case you or someone else has an idea it would be highly appreciated.

**Comment by mpilquist** [comment]
Sure, could you share a little bit more of the type signatures you have to fit things in to?

**Comment by sebastianvoss** [comment]
@mpilquist Great, thanks for your feedback! I tried to cut it down to a sample program. 

Essentially I need to get from a `Stream[F, A]` to `IO[Response[IO]]`. 

The challenge is that I need to tell `Response` which `status` it represents (e.g. `Ok` or `InternalServerError`). To know that I need to `peek` into the `Stream`.

```
import cats.effect.IO
import cats.implicits._
import fs2._
import org.http4s.{HttpRoutes, Response}
import org.http4s.dsl.Http4sDsl

object Sample extends Http4sDsl[IO] {

  val dataFromDoobie: Stream[IO, String] = ???

  val routes: HttpRoutes[IO] = HttpRoutes.of[IO] {
    /* This was my (failed) attempt to first peek into the stream and check if
     * it starts to deliver data. Only in this case I return 200 OK. Otherwise
     * 500 Internal Server Error is returned
     */
    case GET -> Root / "data1" =>
      val data = dataFromDoobie.pull.peek
        .flatMap {
          case Some((_, s)) => Pull.output1(Ok(s))
          case None         => Pull.output1(InternalServerError("error"))
        }
      val response: IO[Response[IO]] = data.stream.compile.lastOrError.flatten
      response

    /* This works but blindly responds with 200 OK. In case of failure
     * the client receives an HTTP chunking error
     */
    case GET -> Root / "data2" =>
      val response: IO[Response[IO]] = Ok(dataFromDoobie)
      response
  }

}
```

**Comment by mpilquist** [comment]
Hm, why does that first attempt fail? It looks okay to me...

**Comment by sebastianvoss** [comment]
@mpilquist I seems the `dataFromDoobie` stream gets finalized when doing `peek`. It results in `PSQLException: This ResultSet is closed`.

**Comment by mpilquist** [comment]
Okay, this reproduces the issue: https://scastie.scala-lang.org/ZzmRMECeRDewUKaHWLq3kA

Note there's still 2 compilations here -- 1 in `myRoute` and another in `serve` (simulating internals of http4s).

The following fixes this by using scope leasing -- it's pretty complicated. We should make this simpler, either in http4s or in fs2, not sure. https://scastie.scala-lang.org/gEEvOvqpSd6NLtRu1Y2oNA

/cc @rossabaker @ChristopherDavenport @SystemFw

**Comment by mpilquist** [comment]
Fixed by #1996
